Use proper perl "signatures" in consoles::VNC

okurz opened the pull request:
Using a command like

```
sed -i "/^use strict;/ {N;s/use strict;/use Mojo::Base -strict, -signatures;/}" $(git ls-files)
for i in $(git grep -l 'use \(strict\|warnings\)'); do grep -q 'use Mojo::Base -strict' $i && sed -i '/use \(strict\|warnings\)/d' $i; done
tools/tidy --only-changed
```

and some manual fixes.

Martchus commented on consoles/serial_screen.pm:
```diff
@@ -168,9 +161,7 @@ An undefined timeout will cause to wait indefinitely. A timeout of 0 means to
-sub do_read
-{
-    my ($self, undef, %args) = @_;
+sub do_read ($self, undef, %args) {
```
This one looks quite strange (include the previous version). It is called like this:

```
        my $read = $self->do_read($buf, max_size => $buflen / 2, timeout => remaining($sttime, $timeout));
        next READ unless (defined($read));
```

So `$buf` is basically ignored. Maybe we should get rid of the parameter in both functions and the invocation.

okurz replied:
For now I changed the second parameter to be `$` to ignore. I will look into your suggestion

Martchus commented on consoles/serial_screen.pm:
```diff
@@ -231,8 +222,7 @@ C<{ matched => 0, string => 'text from the terminal' }>
-    my ($self, $pattern, $timeout) = @_[0 .. 2];
+sub read_until ($self, $pattern, $timeout) {
     my $fd       = $self->{fd_read};
     my %nargs    = @_[3 .. $#_];
```
This line must go into the signature as well, otherwise the function only expects 3 arguments.

okurz replied:
done

Martchus commented on consoles/ssh_screen.pm:
```diff
@@ -25,9 +25,8 @@ has ssh_channel    => undef;
-sub new {
-    my $class = shift;
-    my $self  = bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;
+sub new ($class) {
+    my $self = bless @_ ? @_ > 1 ? {@_} : {%{$_[0]}} : {}, ref $class || $class;
```
I don't think this is going to work. The function will only accept one argument and is `@_` really implicitly shifted?

okurz replied:
Right. Changed that to a second argument `@args`

Martchus reviewed (changes requested):
Hard to tell whether it'll work without checking all invocations. I guess we'll have to risk it but it would make nevertheless sense to run at least some openSUSE tests locally.

cfconrad commented on consoles/serial_screen.pm:
```diff
@@ -166,20 +136,15 @@ sub do_read
+        croak "Failed to read from virtio/svirt serial console char device: $ERRNO" if !defined($read) && !($ERRNO{EAGAIN} || $ERRNO{EWOULDBLOCK});
     }
-    $_[1] = $buffer;
+    $buf = $buffer;
```
We use this called-by-reference, its equal to the `sysread(..., $buffer,...)` not sure how to do this with signatures.
```suggestion
    $_[1] = $buffer; 
```

Martchus replied:
We could require one to specify a scalar ref here. That would make it more obvious that the variable *outside* the function is re-assigned.

okurz replied:
do you have an idea why this test would only fail within OBS and not GHA?

cfconrad replied:
Looks like some GHA glitch it failes: https://github.com/os-autoinst/os-autoinst/runs/4386959135?check_suite_focus=true#step:3:572

perlpunk replied:
Assigning to `$_[...]` willl still work with signatures, as `@_` is still working as before.

There is a dicussion on removing `@_`, though:
https://www.nntp.perl.org/group/perl.perl5.porters/2021/11/msg262052.html

kalikiana commented:
I see a failure, although it doesn't fail CI:

```
8: [20:35:09] ./13-osutils.t ........................... ok     1221 ms ( 0.01 usr  0.00 sys +  1.05 cusr  0.17 csys =  1.23 CPU)
8: 
8:     #   Failed test 'isotovideo exit code'
8:     #   at ./14-isotovideo.t line 128.
8:     #          got: '1'
8:     #     expected: '0'
8: 
8:     #   Failed test 'requested image is published even though the job failed'
8:     #   at ./14-isotovideo.t line 131.
```

okurz commented:
pulled out one part with https://github.com/os-autoinst/os-autoinst/pull/1899

perlpunk commented on consoles/serial_screen.pm:
```diff
@@ -118,19 +104,13 @@ sub normalise_pattern ($pattern, $no_regex) {
         my $hr = shift @$pattern;
         if (@$pattern > 0) {
             my $re = qr/($hr)/;
-            for my $r (@$pattern) {
-                $re .= qr/|($r)/;
-            }
+            $re .= qr/|($_)/ for @$pattern;
```
hm, looks like all this can be simplified (and corrected) by:
```
my $re = join "|", map { "($_)" } @$pattern;
return qr{$re};
```
This way also in the case of one element it will get parens `()`

perlpunk replied:
I wonder why the various patterns should all get their own parens.
`(a|b|c)` should be the same as `(a)|(b)|(c)`, and then the code which uses the regex can just use `$1` instead of `@@LAST_MATCH_START`

okurz replied:
The above does not work for me. Fails with

```
t/10-terminal.t .. 1/?   
#   Failed test '[Child 22536] direct: use array of regexs'
#   at t/10-terminal.t line 366.
#                   '
# 
# 
# Welcome to SUSE Linux Enterprise Server 12 SP2 RC3 (x86_64) - Kernel 4.4.21-65-default (hvc0).
# 
# linux-5rw7'
#     doesn't match '(?^u:(?^u:.*\
# \
# Welcome\ to\ SUSE\ Linux\ Enterprise\ Server\ 12\ SP2\ RC3\ \(x86_64\)\ \-\ Kernel\ 4\.4\.21\-65\-default\ \(hvc0\)\.\
# \
# linux\-5rw7\ login\:\ ))'
# Looks like you failed 1 test of 31.
```

perlpunk replied:
Then the test should be adjusted. The regex is needlessly verbose. But I guess that's for another PR.

perlpunk replied:
Yeah. Somehow I wasn't able to mark the whole block for my comment.

okurz commented:
An openQA fullstack test against this change was also successful.

okurz commented:
Updated with a 100% statement coverage unit test for consoles::vnc_base

Martchus commented on consoles/VNC.pm:
```diff
@@ -122,9 +122,7 @@ my @encodings = (
-sub login {
-    my ($self, $connect_timeout) = @_;
-    $connect_timeout //= 10;
+sub login ($self, $connect_timeout = 10) {
```
This will break. Setting a default via signatures is *not* equivalent to the current default initialization. We call the function like this:

```
$self->{vnc}->login($args->{connect_timeout});
```

If `$args->{connect_timeout}` is `undef` we don't get the default value with your change.

okurz replied:
So what do you suggest? Does

```suggestion
sub login ($self, $connect_timeout = undef) {
    $connect_timeout //= 10;
```

work?

Martchus replied:
Yes, this should work.

okurz replied:
ok, updated the change accordingly

okurz commented:
@Martchus

> Unfortunately coverage for consoles/VNC.pm is missing.

I agree. Though I have the feeling some reporting for that is missing. See e.g. https://app.codecov.io/gh/os-autoinst/os-autoinst/compare/1789/diff?src=pr&el=tree#diff-Y29uc29sZXMvVk5DLnBt, It can't be true that only 5% is covered, considering t/99-full-stack.t. Maybe we need to put some additional `Devel::Cover::report() if Devel::Cover->can('report')` in the code somewhere but I don't know yet where.

Martchus commented:
It should be covered by the fullstack test but that one runs the driver code in a sub process. So yes, `Devel::Cover::report() if Devel::Cover->can('report')` is likely needed somewhere. We already have it in `baseclass.pm` but maybe not in all required places and maybe the condition `Devel::Cover->can('report')` is not true when running the driver process as part of the fullstack test.

okurz commented:
Updated with further small improvements and more tests, e.g. for coverage in consoles::serial_screen. No relevant changes to consoles::VNC coverage however as I still don't know how to approach that.

okurz commented:
After https://github.com/os-autoinst/os-autoinst/pull/1913 I should check again if we actually covered all consoles

okurz commented:
18-qemu-options.t fails, don't understand the specific problem yet. Probably should bisect.

okurz commented:
Split out https://github.com/os-autoinst/os-autoinst/pull/1926 and updated this PR to now only include the single left commit about consoles::VNC. The patch coverage will be still be our automatic acceptance threshold due to background processes not collecting coverage properly. Please review https://github.com/os-autoinst/os-autoinst/pull/1926 first or approve both here if you can
